File: docs/breez-sdk/snippets/rust/src/list_payments.rs
use anyhow::Result;
use breez_sdk_spark::*;

#[allow(dead_code)]
async fn get_payment(sdk: &BreezSdk) -> Result<Payment> {
    // ANCHOR: get-payment
    let payment_id = "<payment id>".to_string();
    let response = sdk.get_payment(GetPaymentRequest { payment_id }).await?;
    let payment = response.payment;
    // ANCHOR_END: get-payment

    Ok(payment)
}

#[allow(dead_code)]
async fn list_payments(sdk: &BreezSdk) -> Result<Vec<Payment>> {
    // ANCHOR: list-payments
    let response = sdk.list_payments(ListPaymentsRequest::default()).await?;
    let payments = response.payments;
    // ANCHOR_END: list-payments

    Ok(payments)
}

#[allow(dead_code)]
async fn list_payments_filtered(sdk: &BreezSdk) -> Result<Vec<Payment>> {
    // ANCHOR: list-payments-filtered
    // Filter by asset (Bitcoin or Token)
    let asset_filter = AssetFilter::Token {
        token_identifier: Some("token_identifier_here".to_string()),
    };
    // To filter by Bitcoin instead:
    // let asset_filter = AssetFilter::Bitcoin;

    let response = sdk
        .list_payments(ListPaymentsRequest {
            // Filter by payment type
            type_filter: Some(vec![PaymentType::Send, PaymentType::Receive]),
            // Filter by status
            status_filter: Some(vec![PaymentStatus::Completed]),
            asset_filter: Some(asset_filter),
            // Time range filters
            from_timestamp: Some(1704067200), // Unix timestamp
            to_timestamp: Some(1735689600),   // Unix timestamp
            // Pagination
            offset: Some(0),
            limit: Some(50),
            // Sort order (true = oldest first, false = newest first)
            sort_ascending: Some(false),
            spark_htlc_status_filter: None,
        })
        .await?;
    let payments = response.payments;
    // ANCHOR_END: list-payments-filtered

    Ok(payments)
}

File: docs/breez-sdk/snippets/go/list_payments.go
package example

import (
	"errors"

	"github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
)

func GetPayment(sdk *breez_sdk_spark.BreezSdk) (*breez_sdk_spark.Payment, error) {
	// ANCHOR: get-payment
	paymentId := "<payment id>"
	request := breez_sdk_spark.GetPaymentRequest{
		PaymentId: paymentId,
	}
	response, err := sdk.GetPayment(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	payment := response.Payment
	// ANCHOR_END: get-payment
	return &payment, nil
}

func ListPayments(sdk *breez_sdk_spark.BreezSdk) (*[]breez_sdk_spark.Payment, error) {
	// ANCHOR: list-payments
	response, err := sdk.ListPayments(breez_sdk_spark.ListPaymentsRequest{})

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	payments := response.Payments
	// ANCHOR_END: list-payments
	return &payments, nil
}

func ListPaymentsFiltered(sdk *breez_sdk_spark.BreezSdk) (*[]breez_sdk_spark.Payment, error) {
	// ANCHOR: list-payments-filtered
	// Filter by asset (Bitcoin or Token)
	tokenIdentifier := "token_identifier_here"
	var assetFilter breez_sdk_spark.AssetFilter = breez_sdk_spark.AssetFilterToken{TokenIdentifier: &tokenIdentifier}
	// To filter by Bitcoin instead:
	// var assetFilter breez_sdk_spark.AssetFilter = breez_sdk_spark.AssetFilterBitcoin

	// Filter options
	typeFilter := []breez_sdk_spark.PaymentType{
		breez_sdk_spark.PaymentTypeSend,
		breez_sdk_spark.PaymentTypeReceive,
	}
	statusFilter := []breez_sdk_spark.PaymentStatus{
		breez_sdk_spark.PaymentStatusCompleted,
	}
	fromTimestamp := uint64(1704067200) // Unix timestamp
	toTimestamp := uint64(1735689600)   // Unix timestamp
	offset := uint32(0)
	limit := uint32(50)
	sortAscending := false

	request := breez_sdk_spark.ListPaymentsRequest{
		TypeFilter:    &typeFilter,    // Filter by payment type
		StatusFilter:  &statusFilter,  // Filter by status
		AssetFilter:   &assetFilter,   // Filter by asset (Bitcoin or Token)
		FromTimestamp: &fromTimestamp, // Time range filters
		ToTimestamp:   &toTimestamp,   // Time range filters
		Offset:        &offset,        // Pagination
		Limit:         &limit,         // Pagination
		SortAscending: &sortAscending, // Sort order (true = oldest first, false = newest first)
	}
	response, err := sdk.ListPayments(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return nil, err
	}

	payments := response.Payments
	// ANCHOR_END: list-payments-filtered
	return &payments, nil
}
